### DISCUSSION TITLE
How to make pulldown menus?

### DISCUSSION THREAD

**Discussion by melodical-me** [sqlitebrowser/sqlitebrowser#2859] [Q&A]
Right now, I am trying to figure out how to create drop down menus, if they are possible. 

Someone over at stackoverflow said I might have to use a lookup table, which I created formy DB, but then I couldnt figure out how to get the actual pulldown menu to show up in the Browse Data tab.

This is an awesome app!  Thank you for sharing!

**Comment by chrisjlocke** [comment]
By dropdown menus, do you mean foreign key lookups, so rather than putting '1' or '2' in your table, you select 'widget' or 'wobble' and it populates the necessary foreign key value of '1' and '2' ?

**Reply by melodical-me** [reply]
That certainly sounds like a fine possibility, especially if you were running a query where you needed to use foreign keys.

But no, that's not really what I meant here.

I simply meant like, suppose you had a lookup table with something like a list of the five major epochs in the history of the globe:

Cenozoic
Paleozoic
Mesozoic
Cretaceous
Eocene

So, like, they are all unique values in and of themselves, and they could in this sense be used as foreign keys. However, I dont need to use them as foreign keys.  I simply would rather just use a mouse to point and click a dropdown menu so I dont have to type all that out every time I want to complete an entry in a table.  Typing could actually lead to more mistakes anyway, since once the computer gets the right spelling it always has the right spelling versus manual re-typing each word in its entirety (which could lead to typos).

So what I am asking seems pretty simple.  Assume I made a table ("Eras") of just those five terms above.  How would I make a pulldown menu to access those terms in an alternate table?

Using a foreign key would also still be useful if the terms were like long paragraphs or entities with other fields or columns, so any solutions there would be welcome, too.  But, really, I am just a lazy, errant typist...  :)

**Reply by chrisjlocke** [reply]
Well, DB4S sort of does this for you already.

When you start typing in a field in the 'browse data' grid, it'll find matching data and suggest it.

![image](https://user-images.githubusercontent.com/7872218/136845020-be1e1660-8dbd-4af5-bf95-14ff73cdbd7c.png)

![image](https://user-images.githubusercontent.com/7872218/136845057-27b9e147-1924-44de-8f8f-ba1a86db6854.png)

If you're entering figures, this can actually be a pain in the butt, but hey.  Each to their own.

What you're after can't be done in DB4S.  You can't create your own selection of items for this menu.

> Using a foreign key would also still be useful if the terms were like long paragraphs or entities with other fields

You've sort of got the idea of foreign keys, as you mention 'typing can lead to mistakes', but they're not used for long paragraphs.  Fields in a table usually relate to one piece of information - city, postcode, sex, height, etc.  If you had five terms (your Eras) you would put them in a table and they'd have IDs, eg, 1=Cenozoic, 2=Paleozoic, etc.  Then instead of typing those, you use the numbers.  DB4S (or more specifically, SQLite) then joins these together.  Here it is in action.

So, create your table, 'epochs'.

![image](https://user-images.githubusercontent.com/7872218/136845799-ac1ff08d-b664-4ee4-98f0-55940c821d41.png)

Two fields.  An id (integer) and the text for that entry.

Now, in the other table (bongos) we simply use 'epochId'

![image](https://user-images.githubusercontent.com/7872218/136845856-c47d0f8c-dd5c-4121-898e-f35898722bac.png)

So bananas has an epochId of 3.  In our lookup table, that is Mesozoic.

Unfortunately, DB4S doesn't help us enter this data.  When browsing the bongos table, it only displays 3, and not Mesozoic.  Also, if I want to change it to Cenozoic, it can't tell me what the number should be.  It's a shame, but hey.  You can configure foreign keys, and DB4S will show you the numbers 1-5, but it won't tell you what they are.  So, it's half-way there.

Using SQL, you can join the two tables together.

![image](https://user-images.githubusercontent.com/7872218/136847140-e4a1b9ed-82a1-4589-9b3d-9053605d9199.png)

You can tidy this up by making a view of just the fields you want.  The downside (isn't there always a downside?) a view is just that - a view - and again, you can't edit the data.  Some here will say you can edit data in a view, but don't go down that road yet.... down there be dragons.

**Reply by chrisjlocke** [reply]
While I'm here, some here will scrinch up their eyes, their toes will curl and they will scream internally (very loud) as I used 'epochId' and not epoch_id or even (shudder) Epoch_id.  Or Epoch_ID.  OK, the main rule in SQL is that field names (and table names) shouldn't have spaces in them, so we can't have 'epoch id'.  Some people love field names to have capital first letters.  Name, Address, City, etc.  Some love camel case.  This is where spaces are removed from words, but each new word has a capital letter, so maryWasALovelyWoman.  Others love underscores - mary_was_a_lovely_woman.  
Here endeth the lesson. 😉😁

**Reply by melodical-me** [reply]
Thank you for your very helpful answer. :) 

Before closing I have a tangential question, though.  

I'm currently a hobbyist programmer.  I lost my job due to covid, and I have been wanting to lean in the direction of programming.  That said, I am still learning my way around developer sites, like github and stackoverflow, etc... 

What I am wondering is, since DB Browser doesnt do what I was asking, it nevertheless seems like it could then be open space for development.  It might not be the biggest and best idea, but - for a beginner - adding a pulldown feature might make for a nice little project.

I dont know if anyone would actually be interested in developing such a feature, but it cant hurt to ask.  I love DB Browser, and I can see lots of possibilities for it.  I just dont know how to assess interest or tap the development process to build such features.

**Reply by chrisjlocke** [reply]
> It might not be the biggest and best idea, but - for a beginner - adding a pulldown feature might make for a nice little project.

I'm still not sure what you're after.  I think I know - so rather than the suggested dropdown of previous entries (as currently), it only shows you a selection of entries from a pre-defined list.  Bear in mind you can only program around the existing controls though.  The dropdown of previous entries may be something in-built in the control itself, so changing this list may not be possible.  I know in .Net some controls have a list of autocomplete entries (eg, the combobox), but the grid doesn't.

**Reply by jvickroy** [reply]
> 
> 
> Well, DB4S sort of does this for you already.
> 
> When you start typing in a field in the 'browse data' grid, it'll find matching data and suggest it.
> 
Unfortunately, this is not working on my platform (up-to-date, Microsoft Windows 10, DB4S 3.12.2).  Any idea what is wrong?  Thanks. /jim
![DB4S-cell-dropdown](https://user-images.githubusercontent.com/9497294/139358676-d7818eae-3a35-4718-ad60-3c31e4674fe6.png)
